SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,421 @@
+struct PointerIdentification {
+  private var walker = PointerIdentificationUseDefWalker()
+
+  typealias AddressOrPointerArgument = Value
+  mutating func getOriginatingAddressOrArgument(_ atp: PointerToAddressInst) -> AddressOrPointerArgument? {
+    walker.start(atp.type)
+    if walker.walkUp(value: atp.operand, path: SmallProjectionPath()) == .abortWalk {
+      return nil
+    }
+    return walker.result
+  }
+
+  private struct PointerIdentificationUseDefWalker : ValueUseDefWalker {
+    private var addrType: Type!
+    private(set) var result: Value?
+
+    mutating func start(_ addrType: Type) {
+      self.addrType = addrType
+      result = nil
+      walkUpCache.clear()
+    }
+
+    mutating func rootDef(value: Value, path: SmallProjectionPath) -> WalkResult {
+      switch value {
+      case is FunctionArgument:
+        self.result = value
+        return .continueWalk
+      case let atp as AddressToPointerInst:
+        if let result = result, atp.operand != result {
+          self.result = nil
+          return .abortWalk
+        }
+
+        if addrType == atp.operand.type, path.isEmpty {
+          self.result = atp.operand
+          return .continueWalk
+        }
+      default:
+        break
+      }
+      self.result = nil
+      return .abortWalk
+    }
+
+    mutating func walkUp(value: Value, path: SmallProjectionPath) -> WalkResult {
+      switch value {
+      case is BlockArgument, is MarkDependenceInst, is CopyValueInst,
+           is StructExtractInst, is TupleExtractInst, is StructInst, is TupleInst,
+           is FunctionArgument, is AddressToPointerInst:
+        return walkUpDefault(value: value, path: path)
+      default:
+        self.result = nil
+        return .abortWalk
+      }
+    }
+
+    var walkUpCache = WalkerCache<Path>()
+  }
+}
+
+
+/// A walker utility that given an address definition computes the `AccessPath`
+/// of the access (the base address and the address projections to the accessed fields) and
+/// the innermost enclosing scope (`begin_access`).
+struct AccessPathWalker {
+  private var walker = AccessPathWalker()
+  mutating func getAccessPath(of address: Value) -> AccessPath? {
+    assert(address.type.isAddress, "Expected address")
+    walker.start()
+    if walker.walkUp(address: address, path: AccessPathWalker.Path()) == .abortWalk {
+      return nil
+    }
+    return walker.result
+  }
+
+  mutating func getAccessPathWithScope(of address: Value) -> (AccessPath?, EnclosingScope?) {
+    let ap = getAccessPath(of: address)
+    return (ap, walker.scope)
+  }
+
+  mutating func getAccessBase(of address: Value) -> AccessBase? {
+    getAccessPath(of: address)?.base
+  }
+
+  mutating func getAccessScope(of address: Value) -> EnclosingScope? {
+    getAccessPathWithScope(of: address).1
+  }
+
+  private struct AccessPathWalker : AddressUseDefWalker {
```

Comment:
Please don't reuse the same name as the outer struct.
How about simply `Walker`?